### PR TITLE
Auto charge / auto income

### LINKED ISSUE #221: needIncomeSelection should be automated
I'm wondering if needIncomeSelection can be automated.

Intuitively, I'd say yes - is there any counter-example?

### LINKED ISSUE #223: Option to allways (or never) accept power chages

### LINKED ISSUE #226: add option to automate brainstone
if enabled, move the brainstone as much as possible

### PR DESCRIPTION
**Pull request by zeitlinger**
# Auto income

When `autoIncome` is `true`, the player will not be asked about the order of gain token and charge power during income phase.

From all possible combinations, the one with least wasted charges is chosen, "most tokens in area3" as tie-breaker.

# Auto Brainstone

When `autoBrainstone` is `true`, the player will not be asked 

- where to place the brainstone - it wil be placed in the highest area
- about the order of gain and charge during leech (with PI) - it will be done using the strategy as auto income

# Itars auto-charge to area 3

When `itarsAutoChargeToArea3` is `true`, the player will not be asked whenever a token is about to move to area 3

# Decline auto-charge unless it's free

When `autoChargePower` is set to `0`, all auto-charges that are not free will be declined.

# Other changes

The player is always asked when an auto-charge would partially wasted.

- Nothing is wasted: normal auto-charge rules apply
- Partially wasted: ask
- Everything wasted: decline

See [charge rules](https://github.com/boardgamers/gaia-engine/blob/ebf7cda16917f7bf1179e2ad58eba9e17217c413/src/auto-charge.ts#L24-L29).

Fixes #226 
Fixes #223 
Fixes #221

### CODE REVIEW

File: wrapper.ts [hunk 1]
```diff
@@ -55,17 +55,38 @@ export function setPlayerMetaData(engine: Engine, player: number, metaData: { na
...
 
-export function setPlayerSettings(engine: Engine, player: number, settings: { autoCharge?: string }) {
+export function setPlayerSettings(
+  engine: Engine,
+  player: number,
+  settings: {
+    autoCharge?: string;
+    autoIncome?: string;
+    autoBrainstone?: string;
+    itarsAutoChargeToArea3?: string;
```
Comment (coyotte508) [inline wrapper.ts:65]: Those can be boolean (checkbox). It's just there's no support for numbers, but there's support for booleans or string enums.

File: wrapper.ts [hunk 2]
```diff
@@ -55,17 +55,38 @@ export function setPlayerMetaData(engine: Engine, player: number, metaData: { na
...
 
   return engine;
 }
 
 export function playerSettings(engine: Engine, player: number) {
   return {
     autoCharge: String(engine.players[player].settings?.autoChargePower ?? 1),
+    autoIncome: String(engine.players[player].settings?.autoIncome),
+    autoBrainstone: String(engine.players[player].settings?.autoBrainstone),
+    itarsAutoChargeToArea3: String(engine.players[player].settings?.itarsAutoChargeToArea3),
```
Comment (coyotte508) [inline wrapper.ts:89]: Here just use  `!!`

```
autoIncome: !!engine.players[player].settings?.autoIncome,
```

**Review by coyotte508 — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -55,6 +55,7 @@
     "eslint-plugin-promise": "^4.2.1",
     "eslint-plugin-standard": "^4.0.1",
     "mocha": "^5.2.0",
+    "nyc": "^15.1.0",
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply package.json:58]: for running code coverage tests - it's not really important - I can remove it.

[the remark this answers is not in the document]
Comment (coyotte508) [reply package.json:58]: It's fine (and maybe good!), just add the corresponding command to the package.json

File: src/available-command.ts [hunk 4]
```diff
@@ -671,6 +669,31 @@ export function possibleLeech(engine: Engine, player: Player) {
...
+    `1t,${lateLeechValue}${Resource.ChargePower}`,
+    new Reward(Math.max(lateLeechValue - 1, 0), Resource.VictoryPoint).toString()
+  );
+
+  if (settings.autoBrainstone) {
+    //leech as much as possible, but avoid waste
+    if (lateLeechValue > earlyLeechValue) {
+      return [lateLeech];
+    }
+    return [earlyLeech];
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply src/available-command.ts]: that turned out to be a good idea, as it found a lot of corner cases :smile:

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (zeitlinger) [reply package.json]: some other time, when I need it again..

**Comment by coyotte508** [comment]
Thank you for the changes! This is going to be really cool, I'm also going to tweak BGS to show the Itars & Taklon settings only when the player has the corresponding Faction.

I will go over the code once again tomorrow evening and do some tests locally, maybe that will bleed a bit into Thursday.

**Review by coyotte508 — CHANGES_REQUESTED** [review]
The main point is making sure the player settings don't affect the move log. To test that, we can run some of the tests with some player settings enabled (and still without) to check that they work with the same move log. There is an example of when it fails in the comment in `income.ts`.

Do not hesitate to break up the PR if you want to continue working on some of the settings while others are ready. Or keep the one PR :)

File: src/income.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  static create(data: PlayerData, settings: Settings, events: Event[]): IncomeSelection {
+    // we need to check if rewards contains Resource.GainToken and Resource.GainPower
+    // player has to select the order
+    const notActivated = events.filter((ev) => !ev.activated);
+
+    const gainTokens = notActivated.filter((ev) => ev.rewards.some((rw) => rw.type === Resource.GainToken));
+    const chargePowers = notActivated.filter((ev) => ev.rewards.some((rw) => rw.type === Resource.ChargePower));
+
+    const auto = gainTokens.length === 0 || chargePowers.length === 0 || settings.autoIncome;
```
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/income.ts:23]: To add to the comment, currently the `engine.autoCharge` function is only called in the wrapper - line `139`. The other automatic code based on settings - `autoIncome` and `autoBrainstone` - should only be called there. Either by editing the wrapper to call `engine.autoBrainstone()` and `engine.autoIncome()` as well, or by modifying/renaming `engine.autoCharge` to `engine.autoMove()` which does the bot move for the player: the auto charging, but also moving the brainstone / choosing the income.

File: src/income.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  applyGainTokens(data, beforeCharge);
+  const waste = applyChargePowers(data, chargePowers);
+  const gainAfterCharge = gainTokens.filter((event) => !beforeCharge.includes(event));
+  applyGainTokens(data, gainAfterCharge);
+  return { waste: waste, power: data.power, events: beforeCharge.concat(chargePowers).concat(gainAfterCharge) };
+}
+
+export function calculateAutoIncome(data: PlayerData, gainTokens: Event[], chargePowers: Event[]): Event[] {
+  const possibleSequences = combinations(gainTokens).map((beforeCharge) =>
+    runIncomeSimulation(data.clone(), beforeCharge, chargePowers, gainTokens)
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply src/income.ts:67]: Yes, this is done in https://github.com/boardgamers/gaia-engine/pull/229/files#diff-d43791db49a9902f2ab7530713f00aaf85be90ef60c61cf3b1e570ebf00bd38cR60

File: src/income.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zeitlinger) [reply src/income.ts:25]: > I may be mistaken, but the feature you added recently

are you referring to taklons special leech?

File: src/income.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/income.ts:75]: Alright! It threw me off a bit. Maybe add a doc comment to `runIncomeSimulation` or just rename the last argument of `runIncomeSimulation` to `allGainTokens`

File: src/income.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/income.ts:25]: Oh never mind - I was mistaken. I thought the current code handled the income for the player if the order `[...gainTokens, ...chargePowers]` fully charged the income. It should be doable with `canFullyChargeDuringIncomePhase()` (in the current code). 

Anyway nevermind the 'Loss of functionality' part.

[the remark this answers is not in the document]
Comment (zeitlinger) [reply src/income.ts]: Addressed the replayability.

Can you add the test to Taklons? - I find the tests that replay entire games hard to follow.

**Comment by zeitlinger** [comment]
> Do not hesitate to break up the PR if you want to continue working on some of the settings while others are ready. Or keep the one PR :)

no, there's no hurry.

File: src/income.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zeitlinger) [reply src/income.ts]: regarding auto move separation:

I finished that part - I hope I got it right.

File: src/income.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  }
+  return maxCharge.events;
+}
+
+export function combinations<T>(t: T[]): T[][] {
+  if (t.length == 0) {
+    return [[]];
+  }
+  return combinations(t.slice(1)).flatMap((value) => [value, value.concat(t[0])]);
+}
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply src/income.ts]: I even tried to use https://github.com/SeregPie/lodash.combinations but that didn't have typescript definitions - and it wasn't very readable.

**Comment by zeitlinger** [comment]
I guess we need to refactor the brainstone handling 

1. asking for brainstone destination should not be a separate step - it should rather be included in the charge options right away
2. this will also avoid the exception in the auto charge - where it asks for brainstone - which breaks the auto brainstone - IIUC

File: src/income.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (coyotte508) [reply src/income.ts]: Thank you!

I'll merge and add the tests & stuff

**Comment by coyotte508** [comment]
Regarding the brainsone in a separate step - it should be possible to do it with the current setup, it's just that `autoBrainstone` and `autoCharge` would need to share some logic (specifically `autoCharge` would need to be aware of the player's `autoBrainstone` setting).

I'm not against refactoring if the current setup proves really impractical, but I think it can be done well enough as is - without making too many breaking changes to the format

**Comment by zeitlinger** [comment]
> Regarding the brainsone in a separate step - it should be possible to do it with the current setup, it's just that `autoBrainstone` and `autoCharge` would need to share some logic (specifically `autoCharge` would need to be aware of the player's `autoBrainstone` setting).
> 
> I'm not against refactoring if the current setup proves really impractical, but I think it can be done well enough as is - without making too many breaking changes to the format of the logs

it was a bad idea - I tried it, but the code because rather messy.
I also convinced myself that the current setup should work :smile:

**Comment by coyotte508** [comment]
It sounds like a challenge!

I just tested locally and was confused when my taklons with autocharge 2pw and auto brainstone didn't leech. I guess no sleep for me today ^^

**Comment by zeitlinger** [comment]
Did you get it to work?

**Comment by coyotte508** [comment]
I think so!

I added some tests in the wrapper. I'll do some more testing before putting it live.
